potassium/potassium.py

```diff
@@ -166,107 +197,67 @@ def wrapper(request):
-
-            thread = Thread(target=task, args=(endpoint, self._gpu_lock, req))
-            thread.start()
-
-            # send task start success message
-            res = make_response({'started': True})
-        else:
-            raise InvalidEndpointTypeException()
-
-        return res
-
-    # WARNING: cover depends on this being called so it should not be changed
-    def _read_event_chan(self) -> bool:
-        """
-        _read_event_chan essentially waits for a background task to finish, 
-        and then returns True
-        """
-        with self._background_task_cv:
-            # wait until the background task is done
-            self._background_task_cv.wait()
-        return True
-
     def _create_flask_app(self):
         flask_app = Flask(__name__)
 
         # ingest into single endpoint and spread out to multiple downstream funcs
         @flask_app.route('/', defaults={'path': ''}, methods=["POST"])
         @flask_app.route('/<path:path>', methods=["POST"])
         def handle(path):
-            with self._sequence_number_lock:
-                self._sequence_number += 1
-
             route = "/" + path
             if route not in self._endpoints:
                 abort(404)
 
             endpoint = self._endpoints[route]
-            return self._handle_generic(endpoint, request)
-        
+            request_id = request.headers.get("X-Banana-Request-Id", None)
+            if request_id is None:
+                request_id = str(uuid.uuid4())
+            try:
+                req = Request(
+                    headers=RequestHeaders(dict(request.headers.items())),
+                    json=request.get_json(),
+                    id=request_id
+                )
+            except:
+                res = make_response()
+                res.status_code = 400
+                return res
+
+            self._event_queue.put((StatusEvent.INFERENCE_REQUEST_RECEIVED,))
+
+            assert self._worker_pool is not None, "Worker pool not initialized"
+            # use an internal id for critical path to prevent user from accidentally
+            # breaking things by sending multiple requests with the same id
+            internal_id = str(uuid.uuid4())
+            if endpoint.type == "handler":
+                self._worker_pool.apply_async(run_worker, args=(endpoint.func, req, internal_id, True))
+                resp = self._response_mailbox.get_response(internal_id)
+
+                flask_response = FlaskResponse(
+                    resp.body,
+                    status=resp.status,
+                    headers=resp.headers
+                )
+            elif endpoint.type == "background":
+                self._worker_pool.apply_async(run_worker, args=(endpoint.func, req, internal_id))
+
+                flask_response = make_response({'started': True})
+            else:
+                raise InvalidEndpointTypeException()
+
+            return flask_response
```

Comment:
there's a lot going on here that could probably be broken down into private methods.